scripts/phrase_context_classifier.py

```diff
@@ -0,0 +1,92 @@
+import torch
+from torch import nn
+import scripts.composition_functions as comp_functions
+import torch.nn.functional as F
+
+
+class PhraseContextClassifier(nn.Module):
+    def __init__(self, embedding_dim, hidden_size, num_layers, forward_hidden_dim, label_nr, dropout_rate):
+        """
+        This class contains the Phrase And Context Classifier. As input the classifier retrieves a sequence of word
+        vectors (= context) and two words of a phrase. The sequence is at first encoded with a bi-directional LSTM.
+        The encoded sequence and the phrase (concatenated word vectors) are then fed through a nonlinear classifier.
+        :param embedding_dim: embedding dimension of a single word
+        :param hidden_size: hidden layer size of the LSTM
+        :param num_layers: if more than one, several LSTM layers are stacked upon each other
+        :param forward_hidden_dim: the hidden layer size of the forward classifier
+        :param label_nr: the number of classes for classification
+        :param dropout_rate: the dropout rate that is applied to the forward classifier
+        """
+        super(PhraseContextClassifier, self).__init__()
+        self._hidden_size = hidden_size
+        self._embedding_dim = embedding_dim
+        self._num_layers = num_layers
+        self._forward_hidden_dim = forward_hidden_dim
+        self._label_nr = label_nr
+        self._dropout_rate = dropout_rate
+        self._lstm = nn.LSTM(self.embedding_dim, hidden_size, num_layers, batch_first=True, bidirectional=True)
+
+        self._hidden_layer = nn.Linear(2 * embedding_dim + hidden_size * 2, forward_hidden_dim)
+        self._output_layer = nn.Linear(forward_hidden_dim, label_nr)
+
+    def forward(self, word1, word2, context, context_lengths, training, device):
+        # context size = batchsize x max len x embedding dim
+
+        # Set initial states
+        # shape = 2 (=bidirectional), batch_size, hidden_size
+        h0 = torch.zeros(self.num_layers * 2, context.size(0), self.hidden_size).to(device)  # 2 for bidirectional LSTM
+        c0 = torch.zeros(self.num_layers * 2, context.size(0), self.hidden_size).to(device)
+
```

Comment:
Hey, ich habe es nochmal nachgelesen. Die beiden ersten states müssen immer wieder neu erstellt werden weil wir jetzt wieder neue sequences anschauen (und sie von vorn bis hinten durchlaufen). Wir gehen davon aus dass unsere sequence unabhängig von einander sind, deswegen lernen wir keine allgemeinen hidden states. Das könnte man machen, wenn man zum Beispiel Language Modelling macht und einen text, Satz für Satz rein steckt. Ich habe aber gesehen dass in der aktuellen LSTM implementation von torch, die initial states nicht explizit gegeben werden müssen. wenn man diese nicht ins LSTM rein gibt, werden alle initial hidden states auf 0 gesetzt wenn das LSTM anfängt einen batch von sequences zu "lesen". Wir können das also weglassen wenn wir das hier eh auch nur mit 0 initialisieren.

Reply:
https://feisky.xyz/machine-learning/apps/torch-rnn/doc/modules.html